src/Test/Hspec/Expectations.hs

```diff
@@ -88,6 +89,23 @@ with_loc(shouldBe, (Show a, Eq a) => a -> a -> Expectation)
 actual `shouldBe` expected = expectTrue ("expected: " ++ show expected ++ "\n but got: " ++ show actual) (actual == expected)
 
 -- |
+-- @actual \`shouldBeNear\` expected@ sets the expectation that @actual@ be a
+-- floating point value near @expected@. If either value is zero we check that
+-- the absolute difference is less than epsilon (1e-15) otherwise we check if
+-- the relative difference is less than epsilon.
+with_loc(shouldBeNear, (Show a, Ord a, Floating a) => a -> a -> Expectation)
+actual `shouldBeNear` expected
+    -- Short circuit if they are actually equal.
+    | actual == expected = reportFail (actual == expected)
+    | actual == 0 || expected == 0 = reportFail (absoluteDifference < epsilon)
+    | otherwise =  reportFail (relativeDifference < epsilon)
```

Comment:
now finally regarding the interesting stuff, the formula, looking at the following example

```
x `shouldBeNear` 1e-300
```
I haven't tried, but what happens if `x` is

1. `0.9e-300`
2. `0`

would (1) fail and (2) pass?

should we always normalize against the expected value instead?

```
actual `shouldBeNear` expected
  | delta < epsilon = return ()
  | otherwise = actual `shouldBe` expected
  where
    epsilon = 1e-15
    absoluteDifference = abs (expected - actual)
    relativeDifference = absoluteDifference / abs expected
    delta
      | expected == 0 = absoluteDifference
      | otherwise = relativeDifference
```